manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+  let r1 = d1 () in
+  let r2 = Domain.join h in
+  assert (not (r1 = 0 && r2 = 0))
+\end{caml_example*}
+
+Under sequential consistency, we would never expect the assertion to fail.
+However, even on x86, which offers much stronger guarantees than the ARM, the
+writes performed at a CPU core are not immediately published to all of the
+other cores. Since "a" and "b" are different memory locations, the reads of "a"
+and "b" may both witness the initial values, leading to the assertion failure.
+
+This behaviour can be explained if a load is allowed to be reordered before a
+preceding store to a different memory location. This reordering can happen due
+to the presence of in-core store-buffers on modern processors. Each core
+effectively has a FIFO buffer of pending writes to avoid the need to block
+while a write completes. The writes to "a" and "b" may be in the store-buffers
+of cores "c1" and "c2" running the domains "d1" and "d2", respectively. The
+reads of "b" and "a" running on the cores "c1" and "c2", respectively, will not
+see the writes if the writes have not propagated from the buffers to the main
+memory.
+
+\section{s:drf_sc}{Data race freedom implies sequential consistency}
+
+The aim of the OCaml relaxed memory model is to precisely describe which orders
+are preserved by the OCaml program. The compiler and the hardware are free to
+optimize the program as long as they respect the ordering guarantees of the
+memory model. While programming directly under the relaxed memory model is
+difficult, the memory model also describes the conditions under which a program
+will only exhibit sequentially consistent behaviours. This guarantee is known
+as \emph{data race freedom implies sequential consistency} (DRF-SC). In this
+section, we shall describe this guarantee. In order to do this, we first need a
+number of definitions.
+
+\subsection{s:atomics}{Memory locations}
+
+OCaml classifies memory locations into \emph{atomic} and \emph{non-atomic}
+locations. Reference cells, array fields and mutable record fields are
+non-atomic memory locations. Immutable objects are non-atomic locations with an
+initialising write but no further updates. Atomic memory locations are those
+that are created using the \stdmoduleref{Atomic} module.
+
+\subsection{s:happens_before}{Happens-before relation}
+
+Let us imagine that the OCaml programs are executed by an abstract machine that
+executes one action at a time, arbitrarily picking one of the available domains
+at each step. We classify actions into two: \emph{inter-domain} and
+\emph{intra-domain}. An inter-domain action is one which can be observed and be
+influenced by actions on other domains. There are several inter-domain actions:
+
+\begin{itemize}
+  \item Reads and writes of atomic and non-atomic locations.
+  \item Spawn and join of domains.
+  \item Operations on mutexes.
+\end{itemize}
+
+\emph{(TODO: Include semaphores and condition variables in the inter-domain
+actions.)}
```

Comment:
TODO, indeed.

Reply:
This should be converted to a latex comment for now.